[Add] support ZeRO Optimizer

yhna940 opened the pull request:
# Title

Colossal AI-based LowLevel zero optimizer with PyTorch interface

# Description

This implementation is a temporary solution to synchronize with Oslo dp. Thank you for your feedback. The existing Colossal AI low-level zero optimizer was introduced almost as is, but the grad scaler was excluded from this integration due to difficulties in aligning the interface. Currently, there is no ddp model wrapper for Oslo, so PyTorch ddp was used for experimentation.

## Usage

```
from torch.nn.parallel import DistributedDataParallel
from oslo.torch.distributed.parallel_context import ParallelContext
from oslo.torch.nn.parallel.data_parallel.zero import ZeroRedundancyOptimizer

parallel_context = ParallelContext.from_torch(data_parallel_size=world_size)
model = ToyModel().to(rank)
ddp_model = DistributedDataParallel(model, device_ids=[rank])
optimizer = optim.SGD(model.parameters(), lr=0.001)
zero_optimizer = ZeroRedundancyOptimizer(optimizer, parallel_context=parallel_context)
zero_optimizer.zero_grad()
outputs = ddp_model(torch.zeros(20, 10).to(rank))
labels = torch.zeros(20, 5).to(rank)
loss_fn(outputs, labels).backward()
zero_optimizer.step()
```


## Limitations

1. Limitations in the original code itself:
1.1. fp16 master weights.
1.2. contiguous gradients.
1.3. ~~cpu offload.~~
1.4. support when some parameters require_grad = False.

2. A new problem arose during this integration process:
2.1. What if there are parameters that do not contribute to the loss function?

## Linked Issues

N/A

## Todo

- [x] Try mixing it with other parallelization methods.

yhna940 commented:
To compare with Colossal ZeRO, I compared the flat parameters of each rank after one step for the following:

1. Colossal ZeRO 1
```
def setup(rank, world_size):
    colossalai.launch(config=dict(), rank=rank, world_size=world_size, port=12346, host='localhost', seed=42)

class ToyModel(nn.Module):
    def __init__(self):
        super(ToyModel, self).__init__()
        self.net1 = nn.Linear(10, 10)
        self.relu = nn.ReLU()
        self.net2 = nn.Linear(10, 5)

    def forward(self, x):
        return self.net2(self.relu(self.net1(x)))


def train(rank, world_size):
    print(f"Running basic DDP example on rank {rank}.")
    setup(rank, world_size)

    # create model and move it to GPU with id rank
    model = ToyModel().to(rank)
    ddp_model = DDP(model, device_ids=[rank])

    loss_fn = nn.MSELoss()
    optimizer = optim.SGD(model.parameters(), lr=0.001)
    zopt = LowLevelZeroOptimizer(
        optimizer,
        min_scale = 1,
    )

    zopt.zero_grad()
    outputs = ddp_model(torch.randn(20, 10))
    labels = torch.randn(20, 5).to(rank)
    zopt.backward(loss_fn(outputs, labels))
    zopt.step()

    print(zopt._fp32_flat_param_groups_of_current_rank)


def main(world_size):
    mp.spawn(train,
             args=(world_size,),
             nprocs=world_size,
             join=True)

if __name__ == "__main__":
    main(2)
```

```
{0: tensor([ 0.2418,  0.2624, -0.0741,  0.2905, -0.0693,  0.0638, -0.1540,  0.1857,
         0.2788, -0.2320,  0.2749,  0.0592,  0.2336,  0.0428,  0.1525, -0.0446,
         0.2438,  0.0467, -0.1476,  0.0806, -0.1457, -0.0371, -0.1284,  0.2098,
        -0.2496, -0.1458, -0.0893, -0.1902,  0.0298, -0.3123,  0.2856, -0.2686,
         0.2442,  0.0526, -0.1027,  0.1954,  0.0493,  0.2555,  0.0346, -0.0997,
         0.0850, -0.0857,  0.1330,  0.2823,  0.1828, -0.1383,  0.1825,  0.0566,
         0.1606, -0.1928, -0.3130, -0.1222, -0.2426,  0.2595,  0.0911,  0.1310,
         0.1000, -0.0055,  0.2475, -0.2247,  0.0199, -0.2159,  0.0975, -0.1089,
         0.0969, -0.0659,  0.2623, -0.1874, -0.1886, -0.1886,  0.2844,  0.1054,
         0.3043, -0.2610, -0.3137, -0.2474, -0.2127,  0.1281,  0.1132,  0.2628,
        -0.1633, -0.2156,  0.1678, -0.1278,  0.1919, -0.0750,  0.1809, -0.2457,
        -0.1596,  0.0964,  0.0669, -0.0806,  0.1885,  0.2150, -0.2293, -0.1688,
         0.2895, -0.1067, -0.1121, -0.3060], device='cuda:0',
       requires_grad=True)}
{0: tensor([ 0.3161,  0.0597,  0.0974, -0.2949, -0.2077, -0.1053,  0.0495, -0.2783,
        -0.1362, -0.1893,  0.0009, -0.1177, -0.0219, -0.2143, -0.2171, -0.1845,
        -0.1083, -0.2496,  0.2651, -0.0628,  0.2721,  0.0985, -0.2678,  0.2188,
        -0.0870, -0.1212, -0.2625, -0.3143,  0.0905, -0.0691,  0.1231, -0.2595,
         0.2348, -0.2321, -0.0546,  0.0661,  0.1633,  0.2553,  0.2881, -0.2507,
         0.0795, -0.1361, -0.0347, -0.2367,  0.2880, -0.2321,  0.1690,  0.1111,
         0.1028, -0.1710, -0.1811,  0.0790, -0.0417, -0.2296,  0.0074, -0.2160,
        -0.2683, -0.1742, -0.2768, -0.2014,  0.2874,  0.0695,  0.0407, -0.2786,
         0.1326], device='cuda:1', requires_grad=True)}
```

2. ZeRO 1 after porting
```
def setup(rank, world_size):
    os.environ['MASTER_ADDR'] = 'localhost'
    os.environ['MASTER_PORT'] = '12345'
    os.environ["RANK"] = str(rank)
    os.environ["LOCAL_RANK"] = str(rank)
    os.environ["WORLD_SIZE"] = str(world_size)
    os.environ["LOCAL_WORLD_SIZE"] = str(world_size)

def cleanup():
    dist.destroy_process_group()

class ToyModel(nn.Module):
    def __init__(self):
        super(ToyModel, self).__init__()
        self.net1 = nn.Linear(10, 10)
        self.relu = nn.ReLU()
        self.net2 = nn.Linear(10, 5)

    def forward(self, x):
        return self.net2(self.relu(self.net1(x)))


def train(rank, world_size):
    print(f"Running basic DDP example on rank {rank}.")
    setup(rank, world_size)
    parallel_context=ParallelContext.from_torch(data_parallel_size=world_size)

    # create model and move it to GPU with id rank
    model = ToyModel().to(rank)
    ddp_model = DDP(model, device_ids=[rank])

    loss_fn = nn.MSELoss()
    optimizer = optim.SGD(model.parameters(), lr=0.001)
    zopt = ZeroRedundancyOptimizer(
        optimizer,
        parallel_context = parallel_context,
    )

    zopt.zero_grad()
    outputs = ddp_model(torch.zeros(20, 10).to(rank))
    labels = torch.zeros(20, 5).to(rank)
    loss_fn(outputs, labels).backward()
    zopt.step()

    print(zopt._fp32_flat_param_groups_of_current_rank)

    cleanup()


def main(world_size):
    mp.spawn(train,
             args=(world_size,),
             nprocs=world_size,
             join=True)

if __name__ == "__main__":
    main(2)
```
```
Running basic DDP example on rank 1.
{0: tensor([ 0.3161,  0.0597,  0.0974, -0.2949, -0.2077, -0.1053,  0.0494, -0.2783,
        -0.1363, -0.1893,  0.0009, -0.1177, -0.0219, -0.2143, -0.2171, -0.1845,
        -0.1082, -0.2496,  0.2651, -0.0628,  0.2721,  0.0985, -0.2678,  0.2188,
        -0.0870, -0.1212, -0.2625, -0.3144,  0.0905, -0.0691,  0.1231, -0.2595,
         0.2348, -0.2321, -0.0546,  0.0661,  0.1633,  0.2553,  0.2881, -0.2507,
         0.0796, -0.1360, -0.0347, -0.2367,  0.2880, -0.2321,  0.1690,  0.1111,
         0.1028, -0.1710, -0.1811,  0.0790, -0.0417, -0.2295,  0.0074, -0.2160,
        -0.2683, -0.1741, -0.2768, -0.2014,  0.2873,  0.0695,  0.0407, -0.2786,
         0.1327], device='cuda:1', requires_grad=True)}
Running basic DDP example on rank 0.
{0: tensor([ 0.2418,  0.2625, -0.0741,  0.2905, -0.0693,  0.0638, -0.1540,  0.1857,
         0.2788, -0.2320,  0.2749,  0.0592,  0.2336,  0.0428,  0.1525, -0.0446,
         0.2438,  0.0467, -0.1476,  0.0806, -0.1457, -0.0371, -0.1284,  0.2098,
        -0.2496, -0.1458, -0.0893, -0.1901,  0.0298, -0.3123,  0.2856, -0.2686,
         0.2441,  0.0526, -0.1027,  0.1954,  0.0493,  0.2555,  0.0346, -0.0997,
         0.0850, -0.0858,  0.1331,  0.2823,  0.1828, -0.1382,  0.1825,  0.0566,
         0.1606, -0.1927, -0.3130, -0.1222, -0.2426,  0.2595,  0.0911,  0.1310,
         0.1000, -0.0055,  0.2475, -0.2247,  0.0199, -0.2158,  0.0975, -0.1089,
         0.0969, -0.0659,  0.2623, -0.1874, -0.1886, -0.1886,  0.2844,  0.1054,
         0.3043, -0.2610, -0.3137, -0.2474, -0.2127,  0.1281,  0.1132,  0.2628,
        -0.1633, -0.2156,  0.1678, -0.1278,  0.1919, -0.0750,  0.1809, -0.2457,
        -0.1596,  0.0964,  0.0669, -0.0806,  0.1885,  0.2150, -0.2293, -0.1688,
         0.2896, -0.1067, -0.1121, -0.3060], device='cuda:0',
       requires_grad=True)}
```

The experiment results show that the porting was successful with almost no parameter difference up to the second decimal place.

yhna940 commented:
The second experiment compared the usage of ZeRO with not using it during training to confirm that ZeRO has no impact on the training in PyTorch native implementation. I conducted a very small-scale experiment using CIFAR100 and ResNet50.

The initial loss log is as follows:
1. Torch Native
```
INFO:root:[1,   100] loss: 4.152, max memory: 643.302
INFO:root:[1,   100] loss: 4.094, max memory: 643.302
INFO:root:[1,   200] loss: 3.443, max memory: 643.302
INFO:root:[1,   200] loss: 3.463, max memory: 643.302
INFO:root:[1,   300] loss: 3.163, max memory: 643.302
INFO:root:[1,   300] loss: 3.156, max memory: 643.302
INFO:root:[1,   400] loss: 2.962, max memory: 643.302
INFO:root:[1,   400] loss: 3.029, max memory: 643.302
INFO:root:[1,   500] loss: 2.933, max memory: 643.302
INFO:root:[1,   500] loss: 2.889, max memory: 643.302
INFO:root:[1,   600] loss: 2.728, max memory: 643.302
INFO:root:[1,   600] loss: 2.804, max memory: 643.302
INFO:root:[1,   700] loss: 2.680, max memory: 643.302
INFO:root:[1,   700] loss: 2.649, max memory: 643.302
INFO:root:[2,   100] loss: 2.578, max memory: 643.302
INFO:root:[2,   100] loss: 2.572, max memory: 643.302
INFO:root:[2,   200] loss: 2.505, max memory: 643.302
INFO:root:[2,   200] loss: 2.516, max memory: 643.302
INFO:root:[2,   300] loss: 2.395, max memory: 643.302
INFO:root:[2,   300] loss: 2.442, max memory: 643.302
INFO:root:[2,   400] loss: 2.416, max memory: 643.302
INFO:root:[2,   400] loss: 2.454, max memory: 643.302
INFO:root:[2,   500] loss: 2.364, max memory: 643.302
INFO:root:[2,   500] loss: 2.353, max memory: 643.302
INFO:root:[2,   600] loss: 2.279, max memory: 643.302
INFO:root:[2,   600] loss: 2.265, max memory: 643.302
INFO:root:[2,   700] loss: 2.259, max memory: 643.302
INFO:root:[2,   700] loss: 2.225, max memory: 643.302
INFO:root:[3,   100] loss: 2.272, max memory: 643.302
INFO:root:[3,   100] loss: 2.223, max memory: 643.302
INFO:root:[3,   200] loss: 2.194, max memory: 643.302
INFO:root:[3,   200] loss: 2.164, max memory: 643.302
INFO:root:[3,   300] loss: 2.158, max memory: 643.302
INFO:root:[3,   300] loss: 2.113, max memory: 643.302
INFO:root:[3,   400] loss: 2.060, max memory: 643.302
INFO:root:[3,   400] loss: 2.189, max memory: 643.302
INFO:root:[3,   500] loss: 2.124, max memory: 643.302
INFO:root:[3,   500] loss: 2.107, max memory: 643.302
INFO:root:[3,   600] loss: 2.008, max memory: 643.302
INFO:root:[3,   600] loss: 2.033, max memory: 643.302
INFO:root:[3,   700] loss: 1.983, max memory: 643.302
INFO:root:[3,   700] loss: 2.056, max memory: 643.302
INFO:root:[4,   100] loss: 1.991, max memory: 643.302
INFO:root:[4,   100] loss: 2.035, max memory: 643.302
INFO:root:[4,   200] loss: 1.945, max memory: 643.302
INFO:root:[4,   200] loss: 2.005, max memory: 643.302
INFO:root:[4,   300] loss: 1.952, max memory: 643.302
INFO:root:[4,   300] loss: 1.944, max memory: 643.302
INFO:root:[4,   400] loss: 1.878, max memory: 643.302
INFO:root:[4,   400] loss: 1.941, max memory: 643.302
INFO:root:[4,   500] loss: 1.939, max memory: 643.302
INFO:root:[4,   500] loss: 1.893, max memory: 643.302
INFO:root:[4,   600] loss: 1.834, max memory: 643.302
INFO:root:[4,   600] loss: 1.917, max memory: 643.302
INFO:root:[4,   700] loss: 1.863, max memory: 643.302
INFO:root:[4,   700] loss: 1.814, max memory: 643.302
```

2. ZeRO1
```
INFO:root:[1,   100] loss: 4.236, max memory: 500.983
INFO:root:[1,   100] loss: 4.189, max memory: 500.983
INFO:root:[1,   200] loss: 3.494, max memory: 500.983
INFO:root:[1,   200] loss: 3.471, max memory: 500.983
INFO:root:[1,   300] loss: 3.218, max memory: 500.983
INFO:root:[1,   300] loss: 3.146, max memory: 500.983
INFO:root:[1,   400] loss: 2.979, max memory: 500.983
INFO:root:[1,   400] loss: 3.025, max memory: 500.983
INFO:root:[1,   500] loss: 2.856, max memory: 500.983
INFO:root:[1,   500] loss: 2.884, max memory: 500.983
INFO:root:[1,   600] loss: 2.698, max memory: 500.983
INFO:root:[1,   600] loss: 2.776, max memory: 500.983
INFO:root:[1,   700] loss: 2.644, max memory: 500.983
INFO:root:[1,   700] loss: 2.648, max memory: 500.983
INFO:root:[2,   100] loss: 2.654, max memory: 500.983
INFO:root:[2,   100] loss: 2.572, max memory: 500.983
INFO:root:[2,   200] loss: 2.517, max memory: 500.983
INFO:root:[2,   200] loss: 2.504, max memory: 500.983
INFO:root:[2,   300] loss: 2.404, max memory: 500.983
INFO:root:[2,   300] loss: 2.416, max memory: 500.983
INFO:root:[2,   400] loss: 2.355, max memory: 500.983
INFO:root:[2,   400] loss: 2.390, max memory: 500.983
INFO:root:[2,   500] loss: 2.334, max memory: 500.983
INFO:root:[2,   500] loss: 2.362, max memory: 500.983
INFO:root:[2,   600] loss: 2.233, max memory: 500.983
INFO:root:[2,   600] loss: 2.275, max memory: 500.983
INFO:root:[2,   700] loss: 2.220, max memory: 500.983
INFO:root:[2,   700] loss: 2.181, max memory: 500.983
INFO:root:[3,   100] loss: 2.208, max memory: 500.983
INFO:root:[3,   100] loss: 2.193, max memory: 500.983
INFO:root:[3,   200] loss: 2.198, max memory: 500.983
INFO:root:[3,   200] loss: 2.210, max memory: 500.983
INFO:root:[3,   300] loss: 2.077, max memory: 500.983
INFO:root:[3,   300] loss: 2.086, max memory: 500.983
INFO:root:[3,   400] loss: 2.108, max memory: 500.983
INFO:root:[3,   400] loss: 2.090, max memory: 500.983
INFO:root:[3,   500] loss: 2.060, max memory: 500.983
INFO:root:[3,   500] loss: 2.046, max memory: 500.983
INFO:root:[3,   600] loss: 2.001, max memory: 500.983
INFO:root:[3,   600] loss: 2.060, max memory: 500.983
INFO:root:[3,   700] loss: 2.008, max memory: 500.983
INFO:root:[3,   700] loss: 1.936, max memory: 500.983
INFO:root:[4,   100] loss: 2.018, max memory: 500.983
INFO:root:[4,   100] loss: 1.985, max memory: 500.983
INFO:root:[4,   200] loss: 1.983, max memory: 500.983
INFO:root:[4,   200] loss: 1.995, max memory: 500.983
INFO:root:[4,   300] loss: 1.904, max memory: 500.983
INFO:root:[4,   300] loss: 1.946, max memory: 500.983
INFO:root:[4,   400] loss: 1.912, max memory: 500.983
INFO:root:[4,   400] loss: 1.954, max memory: 500.983
INFO:root:[4,   500] loss: 1.905, max memory: 500.983
INFO:root:[4,   500] loss: 1.900, max memory: 500.983
INFO:root:[4,   600] loss: 1.844, max memory: 500.983
INFO:root:[4,   600] loss: 1.892, max memory: 500.983
INFO:root:[4,   700] loss: 1.844, max memory: 500.983
INFO:root:[4,   700] loss: 1.770, max memory: 500.983
```

```
def setup(rank, world_size):
    os.environ['MASTER_ADDR'] = 'localhost'
    os.environ['MASTER_PORT'] = '12345'
    os.environ["RANK"] = str(rank)
    os.environ["LOCAL_RANK"] = str(rank)
    os.environ["WORLD_SIZE"] = str(world_size)
    os.environ["LOCAL_WORLD_SIZE"] = str(world_size)
    return ParallelContext.from_torch(data_parallel_size=world_size)

def cleanup():
    dist.destroy_process_group()

class ResNet(nn.Module):
    def __init__(self):
        super(ResNet, self).__init__()
        self.resnet = resnet50(weights=ResNet50_Weights.IMAGENET1K_V2)
        num_ftrs = self.resnet.fc.in_features
        self.resnet.fc = nn.Linear(num_ftrs, 100)

    def forward(self, x):
        x = self.resnet(x)
        return x
        
def build(parallel_context):

    transform_train = transforms.Compose(
        [transforms.RandomCrop(32, padding=4),
         transforms.RandomHorizontalFlip(),
         transforms.ToTensor(),
         transforms.Normalize(mean=[0.485, 0.456, 0.406], std=[0.229, 0.224, 0.225])])
    transform_test = transforms.Compose([
        transforms.ToTensor(),
        transforms.Normalize((0.485, 0.456, 0.406), (0.229, 0.224, 0.225)),
    ])

    trainset = torchvision.datasets.CIFAR100(root='./data', train=True,
                                            download=True, transform=transform_train)
    train_sampler = torch.utils.data.distributed.DistributedSampler(trainset, num_replicas=dist.get_world_size(), rank=dist.get_rank())
    trainloader = torch.utils.data.DataLoader(trainset, batch_size=32, num_workers=4, sampler=train_sampler)

    testset = torchvision.datasets.CIFAR100(root='./data', train=False,
                                           download=True, transform=transform_test)
    testloader = torch.utils.data.DataLoader(testset, batch_size=32,
                                             shuffle=False, num_workers=4)

    net = ResNet().cuda()
    net = DistributedDataParallel(net)
    criterion = nn.CrossEntropyLoss()
    optimizer = AdamW(net.parameters(), lr=0.001, betas=(0.9, 0.999), weight_decay=0.05)
    zopt = ZeroRedundancyOptimizer(
        optimizer,
        parallel_context = parallel_context,
    )

    return net, zopt, criterion, trainloader, testloader
    
def train(rank, world_size):
    parallel_context = setup(rank, world_size)
    net, optimizer, criterion, trainloader, testloader = build(parallel_context)

    net.train()
    for epoch in range(64):
        running_loss = 0.0
        for i, data in enumerate(trainloader, 0):
            inputs, labels = data
            inputs = inputs.cuda()
            labels = labels.cuda()
            optimizer.zero_grad()
            outputs = net(inputs)
            loss = criterion(outputs, labels)
            loss.backward()
            optimizer.step()

            running_loss += loss.item()
            if i % 100 == 99:
                logging.info('[%d, %5d] loss: %.3f, max memory: %.3f' %
                      (epoch + 1, i + 1, running_loss / 100, torch.cuda.max_memory_allocated() / 1024 / 1024))
                running_loss = 0.0

    net.eval()
    correct = 0
    total = 0
    with torch.no_grad():
        for data in testloader:
            images, labels = data
            images = images.cuda()
            outputs = net(images).cpu()
            _, predicted = torch.max(outputs.data, 1)
            total += labels.size(0)
            correct += (predicted == labels).sum().item()

    logging.info('Accuracy of the network on the 10000 test images: %d %%' % (
    100 * correct / total))
```

I confirmed that the Zero we ported this time does not affect the training. Although there is a longer log, I will omit it in this PR.

yhna940 commented:
# Memory Benchmark

1. World Size: 2
2. Input Batch:  
2.1. Dtype: fp32
2.2. Dim: [32, 3, 32, 32]

| Metric              | Naive DP | ZeRO 1   | ZeRO 2   |
|---------------------|----------|----------|----------|
| Torch CUDA max memory | 643.302 MiB | 500.983 MiB | 475.171 MiB |

nijkah commented on oslo/torch/nn/parallel/data_parallel/zero/sharded_optim/_utils.py:
```diff
@@ -0,0 +1,284 @@
+    return tensor
+
+
+def has_inf_or_nan(tensor: torch.Tensor) -> bool:
+    """
```
How about using `torch.isfinite`?

yhna940 replied:
I'm not the author of the original code, so I don't know the exact intention, but when I searched, it seems that there is some kind of floating point exception related to fp16. The intention of this function is probably to have a bigger role in checking if an error occurs when converting fp16 to fp32 and merging with the accumulated gradient (rather than simply checking inf which is nan). I tried to reproduce it, but had difficulties. I'm so sorry. Thanks for the review!

hyunwoongko replied:
I think using the original function is better. Because there are some side effects.

yhna940 replied:
As suggested by you, I used Torch's original function (`torch.isinf` and `torch.isnan`).

nijkah replied:
Then, you may be able to use `torch.isfinite`. in line 137.
https://pytorch.org/docs/stable/generated/torch.isfinite.html

yhna940 replied:
Your suggestion has been reflected. thank you!

hyunwoongko replied:
The original function I mentioned is ColossalAI's function.

yhna940 replied:
I misunderstood your words. Reverted to the original function. thank you.

yhna940 commented on oslo/torch/nn/parallel/data_parallel/zero/sharded_optim/sharded_optim.py:
```diff
@@ -0,0 +1,746 @@
+            self._fp32_flat_param_groups_of_current_rank[
+                group_id
+            ].grad = flat_fp32_avg_grads.to(device)
+            self._grad_store._averaged_gradients[group_id] = []
```
```suggestion
```

yhna940 commented:
# Mixing it with other parallelization methods.
Test to see if it can be used with Oslo's other parallelization techniques.

```python
import logging
import os
import torch
import torch.nn as nn
import torch.optim as optim
import torch.distributed as dist
import torch.multiprocessing as mp
from torch.utils.data import DataLoader, DistributedSampler
from transformers import DistilBertTokenizerFast, DistilBertForSequenceClassification
from datasets import load_dataset

from oslo.torch.distributed.parallel_context import ParallelContext
from oslo.torch.nn.parallel import TensorParallel

def setup(rank, world_size):
    os.environ['MASTER_ADDR'] = 'localhost'
    os.environ['MASTER_PORT'] = '12345'
    os.environ["RANK"] = str(rank)
    os.environ["LOCAL_RANK"] = str(rank)
    os.environ["WORLD_SIZE"] = str(world_size)
    os.environ["LOCAL_WORLD_SIZE"] = str(world_size)
    return ParallelContext.from_torch(tensor_parallel_size=world_size)

# Set up logger
class DistLogger:
    def __init__(self, log_file_path):
        self.logger = logging.getLogger(__name__)
        self.logger.setLevel(logging.DEBUG)
        formatter = logging.Formatter('%(asctime)s - %(levelname)s - %(message)s')
        file_handler = logging.FileHandler(log_file_path)
        file_handler.setLevel(logging.DEBUG)
        file_handler.setFormatter(formatter)
        self.logger.addHandler(file_handler)

    def debug(self, msg, *args, **kwargs):
        if dist.get_rank() == 0:
            self.logger.debug(msg, *args, **kwargs)

    def info(self, msg, *args, **kwargs):
        if dist.get_rank() == 0:
            self.logger.info(msg, *args, **kwargs)

    def warning(self, msg, *args, **kwargs):
        if dist.get_rank() == 0:
            self.logger.warning(msg, *args, **kwargs)

    def error(self, msg, *args, **kwargs):
        if dist.get_rank() == 0:
            self.logger.error(msg, *args, **kwargs)

    def critical(self, msg, *args, **kwargs):
        if dist.get_rank() == 0:
            self.logger.critical(msg, *args, **kwargs)
logger = DistLogger('tp-training.log')

# Load the IMDb dataset
dataset_train = load_dataset('imdb', split='train')
dataset_val = load_dataset('imdb', split='test[:50%]')
dataset_test = load_dataset('imdb', split='test[50%:]')

train_texts = dataset_train['text']
train_labels = dataset_train['label']
val_texts = dataset_val['text']
val_labels = dataset_val['label']
test_texts = dataset_test['text']
test_labels = dataset_test['label']

# Set up tokenizer
tokenizer = DistilBertTokenizerFast.from_pretrained('distilbert-base-uncased')

train_encodings = tokenizer(train_texts, truncation=True, padding=True)
val_encodings = tokenizer(val_texts, truncation=True, padding=True)
test_encodings = tokenizer(test_texts, truncation=True, padding=True)

# Prepare the PyTorch Dataset
class IMDbDataset(torch.utils.data.Dataset):
    def __init__(self, encodings, labels):
        self.encodings = encodings
        self.labels = labels

    def __getitem__(self, idx):
        item = {key: torch.tensor(val[idx]) for key, val in self.encodings.items()}
        item['labels'] = torch.tensor(self.labels[idx])
        return item

    def __len__(self):
        return len(self.labels)


# Set up distributed training
def train(rank, world_size):
    parallel_context = setup(rank, world_size)
    dist.barrier()

    # Set up the model
    device = torch.device('cuda') if torch.cuda.is_available() else torch.device('cpu')
    model = DistilBertForSequenceClassification.from_pretrained('distilbert-base-uncased')
    if torch.cuda.is_available() and torch.cuda.device_count() > 1:
        model = model.to(device)
        model = TensorParallel(model, parallel_context=parallel_context)

    else:
        raise Exception

    # Set up optimizer
    optimizer = optim.AdamW(model.parameters(), lr=5e-5)

    # Load data
    train_dataset = IMDbDataset(train_encodings, train_labels)
    val_dataset = IMDbDataset(val_encodings, val_labels)
    test_dataset = IMDbDataset(test_encodings, test_labels)

    if torch.cuda.is_available() and torch.cuda.device_count() > 1:
        train_sampler = DistributedSampler(train_dataset, num_replicas=world_size, rank=rank)
        train_loader = DataLoader(train_dataset, batch_size=16, sampler=train_sampler)
        val_loader = DataLoader(val_dataset, batch_size=64)
        test_loader = DataLoader(test_dataset, batch_size=64)
    else:
        raise Exception

    num_epochs = 8
    for epoch in range(num_epochs):
        # Training loop
        model.train()
        for idx, batch in enumerate(train_loader):
            optimizer.zero_grad()
            input_ids = batch['input_ids'].to(device)
            attention_mask = batch['attention_mask'].to(device)
            labels = batch['labels'].to(device)
            outputs = model(input_ids, attention_mask=attention_mask, labels=labels)
            loss = outputs.loss
            loss.backward()
            optimizer.step()
            if idx % 100 == 0:
                logger.info(f'Epoch {epoch} Iter {idx}: loss = {loss.item():.4f}, memory = {torch.cuda.memory_allocated() / 1024 / 1024:.4f}')

        # Evaluation loop
        model.eval()
        correct = 0
        total = 0
        for batch in val_loader:
            input_ids = batch['input_ids'].to(device)
            attention_mask = batch['attention_mask'].to(device)
            labels = batch['labels'].to(device)
            outputs = model(input_ids, attention_mask=attention_mask, labels=labels)
            predictions = outputs.logits.argmax(dim=-1)
            correct += (predictions == labels).sum().item()
            total += labels.shape[0]
        accuracy = correct / total
        logger.info(f'Epoch {epoch}: validation accuracy = {accuracy:.4f}')

    # Evaluate the model on the test set
    model.eval()
    correct = 0
    total = 0
    for batch in test_loader:
        input_ids = batch['input_ids'].to(device)
        attention_mask = batch['attention_mask'].to(device)
        labels = batch['labels'].to(device)
        outputs = model(input_ids, attention_mask=attention_mask, labels=labels)
        predictions = outputs.logits.argmax(dim=-1)
        correct += (predictions == labels).sum().item()
        total += labels.shape[0]
    accuracy = correct / total
    logger.info(f'Test accuracy = {accuracy:.4f}')

if __name__ == '__main__':
    world_size = 4
    mp.spawn(train, args=(world_size,), nprocs=world_size, join=True)
```

The test accuracy is the same as 0.9144.

| Metric              | Naive TP | TP + ZeRO 1   | 
|---------------------|----------|----------|
| Torch CUDA max memory | 1023.4453 MiB | 857.5508 MiB |

hyunwoongko commented on oslo/torch/nn/parallel/data_parallel/zero/sharded_optim/_utils.py:
```diff
@@ -0,0 +1,284 @@
+
+        total_norm = total_norm_cuda[0].item() ** (1.0 / norm_type)
+
+    if (
```
Can we use `has_inf_or_nan` here?

yhna940 replied:
I'm sorry, but it seems that this request is difficult to apply. `has_inf_or_nan` takes a tensor as input, converts it to float, and then checks the inf or nan of the elements. This syntax checks the inf and nan of the float `total_norm`.

hyunwoongko commented:
Please run `pre-commit run --all-files`.

yhna940 commented:
> Please run `pre-commit run --all-files`.

I ran a precommit to apply the lint. thank you :)

nijkah commented on oslo/torch/nn/parallel/data_parallel/zero/sharded_optim/_utils.py:
```diff
@@ -0,0 +1,270 @@
+            raise exception
+        return True
+    else:
+        if torch.isinf(fp32_tensor).any() or torch.isnan(fp32_tensor).any():
+            return True
```
```suggestion
        if torch.isfinite(fp32_tensor).any():
            return True
```

yhna940 replied:
```suggestion
        if torch.isinf(fp32_tensor).any() or torch.isnan(fp32_tensor).all():
            return False
        return True
```

nijkah replied:
I think this change is trivial, haha. 😂 Never mind this.

hyunwoongko commented:
Please let me know when you think that this is good to merge.

yhna940 commented:
> Please let me know when you think that this is good to merge.

The basic features of ZoRO seem to have been implemented and verified (stages 1 and 2, communication overlap and cpu offload). Since this optimizer is unlikely to change significantly in the future, it seems safe to merge. If minor changes are needed, I'll make a proper PR later. @hyunwoongko. Thank you 😀
